### PR TITLE
[Block Executor] Parallelize resource group accesses

### PR DESCRIPTION
**Pull request by gelash**
Ready for review but cleaning up tests and adding tests for metadata.

Resource group validation must use entries marked as estimates instead of failing, and on the flip side, validation should occur after execution is finished not only when writing outside of previous write set (vanilla Block-STM rule), but when these (size or metadata) are changed after previous incarnation.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [x] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
Passes resource group tests and did not pass without all proper changes, i.e. confirming prop-tests used to generate problematic corner cases for the size
TODO: adding tests for metadata

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: aptos-move/block-executor/src/executor.rs [hunk 1]
```diff
@@ -124,8 +124,11 @@ where
 
         let mut read_set = sync_view.take_parallel_reads();
 
-        // For tracking whether the recent execution wrote outside of the previous write/delta set.
-        let mut updates_outside = false;
+        // For tracking whether it's required to (re-)validate the suffix of transactions in the block.
+        // May happen, for instance, when the recent execution wrote outside of the previous write/delta
+        // set (vanilla Block-STM rule), or if resource group size or metadata changed from an estimate
+        // (since those resource group validations use rely on estimates).
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/executor.rs:130]: typo: `use rely on` --> just `use` (or just `rely on`)

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 2]
```diff
@@ -361,6 +361,37 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
         }));
     }
 
+    /// Versioned write of metadata at a given resource group key (and version).
```
Comment (georgemitenkov) [inline aptos-move/mvhashmap/src/versioned_data.rs:364]: nit: docstring can say that it "Returns true if the previously stored metadata has changed" or something like that?

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 3]
```diff
@@ -361,6 +361,37 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
...
+                ValueWithLayout::Exchanged(arc_data.clone(), None),
+            )),
+        );
+
+        // Changes versioned metadata that was stored.
+        prev_entry.map_or(true, |entry| -> bool {
+            if let EntryCell::Write(_, existing_v) = &entry.cell {
+                !existing_v
+                    .extract_value_no_layout()
+                    .eq_ignoring_bytes(&*arc_data)
```
Comment (georgemitenkov) [inline aptos-move/mvhashmap/src/versioned_data.rs:388]: nit: I think `as_ref()` can be used instead of `&*`?

File: aptos-move/block-executor/src/executor.rs [hunk 4]
```diff
@@ -279,6 +281,8 @@ where
                 Resource => versioned_cache.data().remove(&k, idx_to_execute),
                 Module => versioned_cache.modules().remove(&k, idx_to_execute),
                 Group => {
+                    needs_suffix_validation = true;
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/executor.rs:284]: I am missing something, why do we need this?

File: aptos-move/block-executor/src/executor.rs [hunk 5]
```diff
@@ -348,8 +352,14 @@ where
                     Resource => versioned_cache.data().mark_estimate(&k, txn_idx),
                     Module => versioned_cache.modules().mark_estimate(&k, txn_idx),
                     Group => {
-                        versioned_cache.data().mark_estimate(&k, txn_idx);
+                        // Validation for both group size and metadata is based on values.
+                        // Execution may wait for estimates.
                         versioned_cache.group_data().mark_estimate(&k, txn_idx);
+
+                        // We are not marking metadata change as estimate, but after
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/executor.rs:359]: I was confused at first by the comment, maybe mention that the metadata lives in data versioned cache and so we do not have to nark it? Otherwise seems only clear when I see the diff with `versioned_cache.data().mark_estimate(&k, txn_idx);` removed 😄

File: aptos-move/mvhashmap/src/versioned_group_data.rs [hunk 6]
```diff
@@ -224,15 +247,22 @@ impl<T: Hash + Clone + Debug + Eq + Serialize, V: TransactionWrite> VersionedGro
...
         // Remove idx updates first, then entries.
-        let idx_update_tags: Vec<T> = self
+        let idx_update_tags: Vec<(T, Option<usize>)> = self
             .idx_to_update
             .remove(&shifted_idx)
-            .map_or(vec![], |map| map.into_inner().into_keys().collect());
+            .map_or(vec![], |map| {
+                map.into_inner()
+                    .into_iter()
+                    .map(|(tag, v)| (tag, v.bytes_len()))
```
Comment (georgemitenkov) [inline aptos-move/mvhashmap/src/versioned_group_data.rs:260]: Why map if we skip the second entry anyway in `for (tag, _) in idx_update_tags.iter() { ...`?

File: aptos-move/mvhashmap/src/versioned_group_data.rs [hunk 7]
```diff
@@ -242,6 +272,10 @@ impl<T: Hash + Clone + Debug + Eq + Serialize, V: TransactionWrite> VersionedGro
             );
         }
 
+        if !replacing && !idx_update_tags.is_empty() {
```
Comment (georgemitenkov) [inline aptos-move/mvhashmap/src/versioned_group_data.rs:275]: I think this is a bit confusing that the caller `write` changes `self.size_changed` outside of this function, but `pub remove` relies on internal changes. Maybe it is better to not have this check here so it is used by write without passing the flag... and have a wrapper remove function here which would be called by `pub remove` and do the extra flag setting?

File: aptos-move/mvhashmap/src/versioned_group_data.rs (hunk 6 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_group_data.rs:260]: we return it, we need to check whether sizes changed.

File: aptos-move/mvhashmap/src/versioned_data.rs (hunk 3 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_data.rs:388]: got rid of eq_ignoring_bytes

File: aptos-move/mvhashmap/src/versioned_group_data.rs (hunk 7 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_group_data.rs:275]: flag was an improvement from the previous iteration and got stuck there - this is even better, thanks!

File: aptos-move/block-executor/src/executor.rs (hunk 4 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:284]: anytime behavior that is observed during speculative execution (and this now includes group metadata and size) changes, we need to trigger suffix re-validation. Previously, we always wait on estimates, so the only way to change behavior is to write outside. Having a write that was there before but not anymore was not an issue - could only cause additional waiting but not an incorrect speculation result.

However, now, the size or metadata might be read, then speculative group update might be removed fully, and without triggering re-validation, some later transaction might end up with the incorrect read result (corresponding to the removed group information that only existed in some incorrect speculative state).

Funny enough, missing this case was exactly the bug I had before when it wasn't passing the tests :) Will add a comment because it is tricky.

**Comment by runtian-zhou** [comment]
Was wondering when those metadata can be changed? IIUC they should stay immutable? Am I missing anything?

File: aptos-move/block-executor/src/proptest_types/types.rs [hunk 8]
```diff
@@ -177,7 +177,7 @@ pub(crate) struct ValueType {
     /// Wrapping the types used for testing to add TransactionWrite trait implementation (below).
     bytes: Option<Bytes>,
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/proptest_types/types.rs:178]: I tried and it's annoying to change, because sometimes we access these fields directly. The whole testing infra here needs some love, for now added a TODO so we remember this improvement.

File: aptos-move/block-executor/src/proptest_types/types.rs [hunk 9]
```diff
@@ -354,21 +355,43 @@ pub(crate) struct MockIncarnation<K, E> {
     /// A vector of keys and corresponding values to be written during mock incarnation execution.
     pub(crate) writes: Vec<(K, ValueType)>,
     pub(crate) group_reads: Vec<(K, u32)>,
-    pub(crate) group_writes: Vec<(K, HashMap<u32, ValueType>)>,
-    /// Keys to query group size for
-    pub(crate) group_sizes: Vec<K>,
+    pub(crate) group_writes: Vec<(K, StateValueMetadata, HashMap<u32, ValueType>)>,
+    /// Keys to query group size for - false is querying size, true is querying metadata.
+    pub(crate) group_queries: Vec<(K, bool)>,
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/proptest_types/types.rs:356]: another one of those, I believe when we execute it we handle them differently (reads go with read results).. all of this needs a thorough refactoring..

File: aptos-move/mvhashmap/src/versioned_group_data.rs [hunk 10]
```diff
@@ -453,11 +487,12 @@ impl<
...
-    /// collected based on the list of recorded tags. If the latest entry at any tag was
-    /// marked as an estimate, a dependency is returned. Note: it would be possible to
-    /// process estimated entry sizes, but would have to mark that if after the re-execution
-    /// the entry size changes, then re-execution must reduce validation idx.
+    /// Returns the sum of latest sizes of all group members (and respective tags), collected
+    /// based on the recorded list of tags. If the latest entry at a tag is marked as estimate
+    /// and the group size has changed between speculative executions then a dependency is
+    /// returned. Otherwise, the size is computed including the sizes of estimated entries.
+    /// This works w. Block-STM, because a validation wave is triggered when any group entry
+    /// size changes after re-execution (incl. size '0', i.e. not matching entries).
```
Comment (igor-aptos) [inline aptos-move/mvhashmap/src/versioned_group_data.rs:495]: what does it mean "size '0', i.e. not matching entries" ?

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 11]
```diff
@@ -361,6 +361,40 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
         }));
     }
 
+    /// Versioned write of metadata at a given resource group key (and version). Returns true
+    /// if the previously stored metadata has changed as observed by later transactions (e.g.
+    /// metadata of a deletion can never be observed by later transactions).
+    pub fn write_metadata(
```
Comment (igor-aptos) [inline aptos-move/mvhashmap/src/versioned_data.rs:367]: stupid question for the future:
- metadata always has basically empty data - empty Bytes, rigth?
- if we change above to compare by value any value that is smaller than some size - this would be covered as well, and would not need a separate function?

**Review by igor-aptos — APPROVED** [review]

**Review by georgemitenkov — APPROVED** [review]

File: aptos-move/mvhashmap/src/versioned_data.rs (hunk 11 above)
Comment (gelash) [reply aptos-move/mvhashmap/src/versioned_data.rs:367]: Correct, I want to change to that with the hints - it might become a bit more nuanced, e.g. might need to have a similar way of not reading if it's been observed to change, even if small, as that leads to doing validations later (after re-execution as opposed to abort).
